temba/orgs/models.py

```
@@ -1114,8 +1114,15 @@ def parse_datetime(self, datetime_string):
     def parse_number(self, decimal_string):
         parsed = None
 
+        if not (isinstance(decimal_string, str)):
+            logger.error(
```

we will log an error if input is not a string type

---

this is something that should never happen.. indicative of a programmer error if it ever does - so I don't think it's appropriate to use logging - especially since we don't want execution to continue. It's more of an assertion and for that we typically throw `ValueError` (tho maybe should be using `assert`...)

---

there is about 10 tests that use `set_field` with a non-string object and that would tripwire this check ... for example, https://github.com/nyaruka/rapidpro/blob/9968b868d6c9a872df6984f040ae17c00000a2cb/temba/campaigns/tests.py#L1284

we probably should check if it 'quacks' like a number, parse it as Decimal, and silently ignore any other types

---

The point is that any use of a Python `float` should be considered programmer error - not just something we log. I'm fine with this method only accepting `Decimal`/`datetime`/`str` (or reworking things so it only accepts `str`).

---

`parse_number` is only used in `Contact.serialize_field` (https://github.com/nyaruka/rapidpro/blob/689bc0f8e81f06f724c6403c1e98a77a97459f21/temba/contacts/models.py#L919) and in that method we parse *everything* as *everything* ... a number is still going to be parsed as: **text**, **datetime** and **location** types

I don't think we should raise exceptions and break the execution when parsing input that is going to be stored as a number

---

Correct me if I'm wrong, but I think there are only 2 places where `set_field` or `set_fields` are called now - the Edit Custom Fields view and the API. In both those cases, the only value types that should be passed in are the JSONy types: `str` and `Decimal` (we should check the API deserializes JSON numbers as `Decimal`).

So what I think we should be avoiding here is creating code paths which are only used by tests and don't reflect actual usage. And in the case of `float`, testing a code path which we should consider a programmer error. A contact field value should never be allowed to become a `float`.

The fact that we have some unit tests which still do things like passing `float` and `datetime` instances to `set_field` isn't an argument for supporting those as valid code paths. It's an argument for updating those unit tests.

Anyways we need to get this merged so maybe @nicpottier @ericnewcomer can weigh in and tell me if I'm being pedantic here.

---

it's probably going to be better in the long run if we decide to define stricter contracts for `parse_number` and `parse_datetime` methods, and **ONLY** allow string input values and raise exceptions if input is not a string (and make sure that exceptions bubble up)

`parse_datetime` was lax about the input value from the initial commit, probably because it was convenient at the time (https://github.com/nyaruka/rapidpro/blob/issue_914/temba/orgs/models.py#L1109)